lib/src/pool/pool_impl.dart

```diff
@@ -347,3 +371,23 @@ class _PoolStatement implements Statement {
     return _underlying.run(parameters, timeout: timeout);
   }
 }
+
+class _PoolConnectionResource {
+  final pool.PoolResource _poolResource;
+  bool _released = false;
+  _PoolConnection? _connection;
+
+  _PoolConnectionResource(this._poolResource);
+
+  /// Same as [PoolResource.release], but it doesn't throw if the resource has
+  /// already been released.
+  /// We might call release outside the [withConnection] block, like when force
```

Comment:
Why would we call release outside the `withConnection` block?

I think we shouldn't need a separate `_PoolConnectionResource`, rather track the same state inside `_PoolConnection`, it seems to me that it should be simpler that way.

Reply:
That's because when force closing the pool, _semaphore.close() waits for all the resources to finish, so it would be waiting for the queries.

I mainly created the wrapper because release can't be called twice, then I added the connection because it was in the same scope/context.

Maybe we can keep the wrapper class and remove the connection as a field. What do you think?

Reply:
> I mainly created the wrapper because release can't be called twice, then I added the connection because it was in the same scope/context.

> Maybe we can keep the wrapper class and remove the connection as a field. What do you think?

Why not merge the two? I really don't see why you couldn't add it to `_PoolConnection`, but maybe I should attempt to implement it myself and see how it feels. Hence my suggestion to land the parts that we agree on and treat the pool as a separate goal.

Reply:
Yes, I'll split the PR.

I originally also thought about adding it to _PgConnection but then I saw that the list of connections is kept separate and I think there is not a 1-1 mapping in the life cycle of the collection and the pool resource. Also, those would be more core changes in the library, that I wouldn't feel comfortable changing too much. But I agree there is room for experimenting there

Reply:
For splitting the PR, should I make the force flag in the pool noop, or should it throw when provided as true?

Reply:
> I think there is not a 1-1 mapping in the life cycle of the collection and the pool resource.

Yeah, I need to review that part of the code, as I have very little memory on how it is done.

> should I make the force flag in the pool noop

Is there any drawback to ignore the flag? I think it should be fine just being a noop, maybe add a TODO to implement it.